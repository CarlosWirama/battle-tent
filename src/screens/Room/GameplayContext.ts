import React, { Dispatch, SetStateAction } from 'react';
import BattlingPokemon from '../../types/BattlingPokemon';
import Pokemon, { Move } from '../../types/Pokemon';
import Player from '../../types/Player';
import { Parties, InboundEventParams } from '../../api/base';

export interface GameplayContextValue {
  opponent: Player | undefined;
  myTurn: boolean;
  changeTurn: (battleState: InboundEventParams['TurnChanged']) => void;
  availablePokemon: Pokemon[];
  availableMoves: Move[][];
  setAvailableMoves: React.Dispatch<React.SetStateAction<Move[][]>>;
  party: BattlingPokemon[];
  opponentParty: BattlingPokemon[];
  updateParties: (parties: Parties) => void;
  energy: number;
  maxEnergy: number;
  setEnergy: Dispatch<SetStateAction<number>>;
}

export default React.createContext<GameplayContextValue>({
  opponent: undefined,
  myTurn: false,
  changeTurn: () => { },
  availablePokemon: [],
  availableMoves: [],
  setAvailableMoves: () => { },
  party: [],
  opponentParty: [],
  updateParties: () => { },
  energy: 0,
  maxEnergy: 1,
  setEnergy: () => { },
});
