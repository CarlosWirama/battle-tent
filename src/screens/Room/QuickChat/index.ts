export { default as QuickChatPanel } from './QuickChatPanel';
export { PlayerChatBalloon, EnemyChatBalloon } from './ChatBalloon.styled';
