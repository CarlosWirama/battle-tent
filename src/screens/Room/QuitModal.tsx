import React from 'react';
import { Button } from '../../components/basics';
import Modal, { ModalProps } from '../../components/Modal';

export default function QuitModal({
  shown,
  onClose,
  onQuit,
}: Omit<ModalProps, 'children'> & { onQuit: () => void }) {
  return (
    <Modal shown={shown} onClose={onClose}>
      Do you really wish to quit?
      <Button onClick={onQuit}>Yes</Button>
      <Button onClick={onClose}>No</Button>
    </Modal>
  );
}
