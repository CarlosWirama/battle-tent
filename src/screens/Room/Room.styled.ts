import styled from 'styled-components';
import { Centered } from '../../components/basics';
import TrainerImage from '../../assets/images/trainer/Red.png';
import { GymLocation, getLocationBackground } from '../../constants/gym';

export const LayoutContainer = styled(Centered) <{ location: GymLocation }>`
  background-color: #0c6479;
  background-image: linear-gradient(141deg,#1294a2 35%,#0a4861 75%);
  background-image: url(${props => getLocationBackground(props.location)});
  background-size: cover;
  background-position: center;
  justify-content: flex-start;
  height: 100vh;
  color: white;
`;

export const TopArea = styled.header`
  display: flex;
  justify-self: flex-start;
  justify-content: space-between;
  width: 100%;
  box-sizing: border-box;
  padding: 16px;
`;
export const MainArea = styled.main`
  display: flex;
  flex-direction: column;
  flex: 1;
  width: 100%;
  position: relative;
`;

export const BottomArea = styled.footer`
  position: absolute;
  bottom: 0;
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  width: 100%;
  box-sizing: border-box;
  padding: 16px;
`;

export const TrainerAvatar = styled.div`
  display: inline-block;
  height: 32px;
  width: 32px;
  border-radius: 50%;
  border: 2px solid white;
  padding: 16px;
  z-index: 1;
  
  // set default image
  background-color: #bbb;
  position: relative;
  overflow: hidden;
  &::after {
    content: "";
    position: absolute;
    top: 2px;
    left: 17px;
    width: 100%;
    height: 100%;
    background-image: url(${TrainerImage});
    background-size: 36px;
    background-repeat: no-repeat;
    filter: contrast(0%) grayscale(1) brightness(0.5);
  }
`;
export const OpponentInfo = styled.div`
  display: flex;
  justify-content: flex-start;
`;
