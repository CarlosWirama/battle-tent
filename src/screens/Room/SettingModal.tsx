import React, { useState } from 'react';
import styled from 'styled-components';
import { Button, Slider } from '../../components/basics';
import Modal, { ModalProps } from '../../components/Modal';
import audio from '../../audio';

export default function SettingModal({
  shown,
  onClose,
  onQuit,
}: Omit<ModalProps, 'children'> & { onQuit: () => void }) {
  const [state, setState] = useState(State.Setting);
  return (
    <Modal shown={shown} onClose={onClose}>
      {state === State.Setting ?
        <SettingLayout onQuitClick={() => setState(State.Quit)} />
        :
        <QuitLayout onCancel={() => setState(State.Setting)} onQuit={onQuit} />
      }
    </Modal>
  );
}

function SettingLayout({ onQuitClick }: { onQuitClick: () => void }) {
  return (
    <SettingLayoutContainer>
      <SoundSettings>
        <div>SFX</div>
        <Slider
          defaultValue={audio.sfxVolume}
          onChangeCommitted={(e, value) => {
            if (Array.isArray(value)) throw value;
            audio.setSfxVolume(value)
          }}
        />
        <div>BGM</div>
        <Slider
          defaultValue={audio.bgmVolume}
          onChange={(_, value) => audio.setBgmVolume(value as number, false)}
        />
      </SoundSettings>
      <Button onClick={onQuitClick}>Quit</Button>
    </SettingLayoutContainer>
  );
}

function QuitLayout({ onQuit, onCancel }: {
  onQuit: () => void;
  onCancel: () => void;
}) {
  return (
    <>
      Do you really wish to quit?
      <Button onClick={onQuit}>Yes</Button>
      <Button onClick={onCancel}>No</Button>
    </>
  );
}

enum State {
  Setting,
  Quit,
}

const SettingLayoutContainer = styled.div`
  font-family: HeadTextFont;
  font-size: 24px;
  color: #874a0c;
`;

const SoundSettings = styled.div`
  display: grid;
  grid-template-columns: 1fr 4fr;
`;
