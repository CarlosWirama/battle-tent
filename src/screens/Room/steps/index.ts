export { default as ChooseParty } from './ChooseParty';
export { default as Battle } from './Battle';
export type { ChoosePartyProps } from './ChooseParty';
