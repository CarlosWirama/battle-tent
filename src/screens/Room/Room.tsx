import React, { useState, useEffect, useContext, useCallback } from 'react';
import { Button, IconButton } from '../../components/basics';
import * as Steps from './steps';
import SettingModal from './SettingModal';
import {
  emitLeaveRoom,
  subscribePlayerJoinedTheRoom,
  subscribePlayerLeftTheRoom,
  subscribeLeftTheRoom,
} from '../../api';
import Player from '../../types/Player';
import { NavigationContext, ScreenState } from '../../navigation';
import { QuickChatPanel } from './QuickChat';
import {
  LayoutContainer,
  TopArea,
  MainArea,
  BottomArea,
  OpponentInfo,
} from './Room.styled';
import { find } from 'ramda';
import { PlayerContext } from '../../auth';
import GameplayContext, { GameplayContextValue } from './GameplayContext';
import BattlingPokemon from '../../types/BattlingPokemon';
import Pokemon, { Move } from '../../types/Pokemon';
import { Parties } from '../../api/base';
import Modal from '../../components/Modal';
import SettingIcon from '../../assets/images/ui/setting.png';
import { GymLocation, getLocationBgm } from '../../constants/location';
import audio from '../../audio';
import { SmallAvatar } from '../../components/Avatar';

enum RoomStep {
  ChooseParty,
  Battle,
}

export default function Room() {
  /** GameplayContexts */
  const [availablePokemon, setAvailablePokemon] = useState<Pokemon[]>([]);
  const [opponent, setOpponent] = useState<Player | undefined>(undefined);

  const [party, setParty] = useState<BattlingPokemon[]>([]);
  const [opponentParty, setOpponentParty] = useState<BattlingPokemon[]>([]);
  const [player] = useContext(PlayerContext);

  function updateParties(parties: Parties) {
    setParty(parties[player.id]);
    opponent && setOpponentParty(parties[opponent.id]);
  }

  const [myTurn, setMyTurn] = useState(false);

  const [energy, setEnergy] = useState(0);
  const [maxEnergy, setMaxEnergy] = useState(1);
  const [availableMoves, setAvailableMoves] = useState<Move[][]>([]);

  const changeTurn: GameplayContextValue['changeTurn'] = ({ my_turn, energy, moves }) => {
    setMyTurn(my_turn);
    setEnergy(my_turn ? energy : 0);
    setMaxEnergy(my_turn ? energy : 1);
    setAvailableMoves(my_turn ? moves : []);
  };
  const gameplayContextValue: GameplayContextValue = {
    opponent,
    myTurn, changeTurn,
    availablePokemon, availableMoves, setAvailableMoves,
    party, opponentParty, updateParties,
    energy, maxEnergy, setEnergy,
  };
  /** End of GameplayContexts */

  const [playerLeftModalShowns, setPlayerLeftModalShowns] = useState(false);
  const [settingModalShown, setSettingModalShown] = useState(false);
  const [activeStep, setActiveStep] = useState(RoomStep.ChooseParty);
  const [location, setLocation] = useState(GymLocation.None);
  const navigate = useContext(NavigationContext);

  const backToLobby = useCallback(
    () => {
      emitLeaveRoom();
      navigate(ScreenState.Lobby);
    },
    [navigate],
  );

  // subscriptions
  useEffect(function subscribe() {
    const sPlayerJoinedTheRoom = subscribePlayerJoinedTheRoom(({
      players,
      availablePokemon,
      gymBattleIndex,
    }) => {
      setLocation(gymBattleIndex);
      const checkMatchingPlayer = ({ id }: Player) => id !== player?.id;
      const findOpponent = find(checkMatchingPlayer);
      const opponent = findOpponent(players);
      opponent && setOpponent(opponent);
      setAvailablePokemon(availablePokemon);
    });

    const sPlayerLeftTheRoom = subscribePlayerLeftTheRoom(({ name }) => {
      setPlayerLeftModalShowns(true);
    });

    const sLeftTheRoom = subscribeLeftTheRoom(backToLobby);

    return function unsubscribe() {
      sPlayerJoinedTheRoom.off();
      sPlayerLeftTheRoom.off();
      sLeftTheRoom.off();
    }
  }, [navigate, player, backToLobby]);

  function openSettingModal() {
    setSettingModalShown(true);
  }
  function closeSettingModal() {
    setSettingModalShown(false);
  }

  const goToBattleStep = () => {
    setActiveStep(RoomStep.Battle);
    audio.playBgm(getLocationBgm(location), { delay: 0 });
  }

  return (
    <LayoutContainer location={location}>
      <TopArea>
        <IconButton aria-label="setting" onClick={openSettingModal}>
          <img alt="setting" src={SettingIcon} width="28px" />
        </IconButton>
        <OpponentInfo>
          <div>{opponent && opponent.name}</div>
          <SmallAvatar code={opponent?.avatar} />
        </OpponentInfo>
      </TopArea>
      <MainArea>
        <GameplayContext.Provider value={gameplayContextValue}>
          {activeStep === RoomStep.ChooseParty ? (
            <Steps.ChooseParty onFinish={goToBattleStep} />
          ) : activeStep === RoomStep.Battle ? (
            <Steps.Battle />
          ) : null}
        </GameplayContext.Provider>
      </MainArea>
      <BottomArea>
        <SmallAvatar code={player.avatar} />
        <QuickChatPanel />
      </BottomArea>
      <SettingModal
        shown={settingModalShown}
        onClose={closeSettingModal}
        onQuit={backToLobby}
      />
      <Modal shown={playerLeftModalShowns} onClose={backToLobby}>
        The opponent left the room!
          <Button onClick={backToLobby}>OK</Button>
      </Modal>
    </LayoutContainer>
  );
}
