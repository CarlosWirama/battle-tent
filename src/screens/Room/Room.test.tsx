import React from 'react';
import { render } from '@testing-library/react';
import Room from './Room';

describe('Room screen', () => {
  it('should render', () => {
    const { asFragment } = render(<Room />);
    expect(asFragment()).toMatchSnapshot();
  });
});
