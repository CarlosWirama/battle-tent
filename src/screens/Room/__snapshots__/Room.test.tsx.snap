// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room screen should render 1`] = `
<DocumentFragment>
  <div
    class="sc-AxhCb sc-fznAgC hrhglL"
  >
    <header
      class="sc-fznLPX eJtHlj"
    >
      <button
        aria-label="setting"
        class="MuiButtonBase-root MuiIconButton-root sc-AxiKw bOTXee"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <img
            alt="setting"
            src="setting.png"
            width="28px"
          />
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="sc-fzqAui qbhzR"
      >
        <div
          class="sc-fzqMdD hZJXXb"
        />
        <div
          class="sc-fzowVh dpZlzW"
        >
          <div
            class="sc-fzoJus fNhbUc"
            src="red.png"
          />
        </div>
      </div>
    </header>
    <main
      class="sc-fzoaKM ikgUwX"
    >
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </main>
    <footer
      class="sc-fzomuh kxRxuq"
    >
      <div
        class="sc-fzowVh dpZlzW"
      >
        <div
          class="sc-fzoJus fNhbUc"
          src="red.png"
        />
      </div>
      <div
        class="sc-fzomME fLFJFI"
      >
        <div
          class="sc-fzqMAW ipguxq"
        >
          <div
            class="sc-fzoydu gHXmzE"
          >
            OK!
          </div>
          <div
            class="sc-fzoydu gHXmzE"
          >
            good game!
          </div>
          <div
            class="sc-fzoydu gHXmzE"
          >
            well played!
          </div>
          <div
            class="sc-fzoydu gHXmzE"
          >
            prepare!
          </div>
          <div
            class="sc-fzoydu gHXmzE"
          >
            take this!
          </div>
          <div
            class="sc-fzoydu gHXmzE"
          >
            stop it!
          </div>
          <div
            class="sc-fzoydu gHXmzE"
          >
            enough!
          </div>
          <div
            class="sc-fzoydu gHXmzE"
          >
            rematch!
          </div>
          <div
            class="sc-fzoydu gHXmzE"
          >
            thank you!
          </div>
        </div>
        <div
          class="sc-fzoJMP sc-fzoYHE BuMHa"
        />
        <button
          class="MuiButtonBase-root MuiIconButton-root sc-AxiKw bOTXee sc-fzqAbL dESkkI"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zM9 11H7V9h2v2zm4 0h-2V9h2v2zm4 0h-2V9h2v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="sc-fzoJMP sc-fzpkqZ fwrEcb"
        />
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
